src/System.Memory/src/System/Buffers/SequenceReader.cs

```diff
@@ -120,6 +120,55 @@ public SequenceReader(ReadOnlySequence<T> sequence)
             }
         }
 
+        /// <summary>
+        /// Peeks at the next value at specific offset without advancing the reader.
+        /// </summary>
+        /// <param name="offset">The offset from current position.</param>
+        /// <param name="value">The next value or default if at the end.</param>
+        /// <returns>False if at the end of the reader.</returns>
+        public readonly bool TryPeek(long offset, out T value)
+        {
+            if (offset < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException_OffsetOutOfRange();
+
+            // if we've data and offset is not out of bounds
+            if (_moreData && Remaining > offset)
+            {
+                int currentIndex = CurrentSpanIndex;
+                long currentOffset = offset;
+                ReadOnlySpan<T> currentSpan = CurrentSpan;
+                SequencePosition nextPosition = _nextPosition;
+
+                // if offset doesn't fall inside current segment move to next until we found correct one
+                while ((currentIndex + currentOffset) > currentSpan.Length - 1)
+                {
+                    // subtract current non consumed data
+                    currentOffset -= currentSpan.Length - currentIndex;
+
+                    // move index to start of next span
+                    currentIndex = 0;
+
+                    // get next segment
+                    if (Sequence.TryGet(ref nextPosition, out ReadOnlyMemory<T> memory, true))
```

Comment:
@JeremyKuhne I'm pretty new to this apis but are interesting...does it make sense re-use public api or the idea was access to internals?

Reply:
I would stackalloc a single item and use the internal `TryCopyToMultiSegment` when we're not in the current segment.

Reply:
ping @JeremyKuhne

Reply:
Sorry, being a bit mental. :) Our existing loop isn't usable as is.

Reply:
no worries 😄